src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,337 @@
+from typing import Any, Dict, List, Optional, Tuple, Union
+from abc import ABC, abstractmethod
+from django.http import QueryDict
+from sentry.utils.snuba import raw_query
+from .dataset import Dataset
+from sentry_relay import DataCategory
+
+from sentry.snuba.sessions_v2 import (
+    get_constrained_date_range,
+    massage_sessions_result,
+    SimpleGroupBy,
+    InvalidField,
+)
+from sentry.utils.outcomes import Outcome
+from datetime import datetime
+from sentry.search.utils import InvalidQuery
+
+
+"""
+The new Outcomes API defines a "metrics"-like interface which is can be used in
+a similar way to "sessions" and "discover"
+# "Metrics"
+We have 2 "metrics" that we can query:
+
+- `quantity` (counter): The relevant stat based on category:
+        standard event: # of outcomes
+        session: number of events within the session
+        attachment: quantity in bytes of attachments
+- `times_seen` (counter): The number of outcomes that occurred --
+        differs from quantity in that:
+            it would count # of attachments, and sessions
+            if we ever aggregate multiple events into one outcome,
+            this would be used to get the number of outcomes
+
+# Tags / Grouping
+
+The Outcomes data can be grouped / filtered
+by these fields
+- `project`
+- `outcome`
+- `reason`
+- `category`
+```
+"""
+
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def extract_from_row(
+        self, row: Dict[str, Union[Any, int]], group: Optional[Dict[str, Any]] = None
+    ) -> int:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(
+        self, row: Dict[str, Union[Any, int]], group: Optional[Dict[str, Any]] = None
```

Comment:
have to do this weird Union to get mypi to like having this function returning an int. wonder what the best way to to this is.

Reply:
fixed, went ahead and just typecasting the return value to `int`, which is essentially a noop here. h/t @RyanSkonnord